client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,317 @@
+
+use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to only events with the key prefixes for :extrinsic_index & system::Account.
+const DEFAULT_STORAGE_KEYS: &'static str =
+	"3a65787472696e7369635f696e646578,26aa394eea5630e07c48ae0c9558cef7b99d880ec681799c0cf30e8886371da9";
+/// Maximal payload accepted by RPC servers. // TODO maybe import from sc_rpc_server
+const MAX_PAYLOAD: usize = 15 * 1024 * 1024;
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Mutex<HashMap<Id, SpanDatum>>,
+	events: Mutex<Vec<TraceEvent>>,
+	timestamp: Instant,
+}
+
+impl BlockSubscriber {
+	fn new(
+		targets: &str,
+		spans: Mutex<HashMap<Id, SpanDatum>>,
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
+			.split(',')
+			.map(|s| crate::parse_target(s))
+			.collect();
+		// Ensure that WASM traces are always enabled
+		// Filtering happens when decoding the actual target / level
+		targets.push((WASM_TRACE_IDENTIFIER.to_owned(), Level::TRACE));
+		BlockSubscriber {
+			targets,
+			next_id,
+			current_span: CurrentSpan::default(),
+			spans,
+			events,
+			timestamp: Instant::now(),
+		}
+	}
+}
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				return true;
+			}
+		}
+		false
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+		let id = Id::from_u64(id);
+		let mut values = Values::default();
+		attrs.record(&mut values);
+		let parent_id = attrs.parent().cloned()
+			.or_else(|| self.current_span.id());
+		let span = SpanDatum {
+			id: id.clone(),
+			parent_id,
+			name: attrs.metadata().name().to_owned(),
+			target: attrs.metadata().target().to_owned(),
+			level: *attrs.metadata().level(),
+			line: attrs.metadata().line().unwrap_or(0),
+			start_time: Instant::now(),
+			values,
+			overall_time: Default::default()
```

Comment:
@mattrutherford - is this getting used anymore?

Reply:
Some of these fields are used in other places in Substrate, but ultimately not in BlockSubscriber